source/cl/test/UnitCL/include/cl_intel_unified_shared_memory.h

```diff
@@ -51,11 +60,116 @@ struct cl_intel_unified_shared_memory_Test : public virtual ucl::ContextTest {
 #undef GET_EXTENSION_ADDRESS
   }
 
+  void TearDown() override {
+    if (device_ptr) {
+      cl_int err = clMemBlockingFreeINTEL(context, device_ptr);
+      EXPECT_SUCCESS(err);
+      device_ptr = nullptr;
+    }
+
+    if (shared_ptr) {
+      cl_int err = clMemBlockingFreeINTEL(context, shared_ptr);
+      EXPECT_SUCCESS(err);
+      shared_ptr = nullptr;
+    }
+
+    if (host_ptr) {
+      cl_int err = clMemBlockingFreeINTEL(context, host_ptr);
+      EXPECT_SUCCESS(err);
+      host_ptr = nullptr;
+    }
+
+    ucl::ContextTest::TearDown();
+  }
+
   static void *getPointerOffset(void *ptr, size_t offset) {
     cl_char *byte_ptr = reinterpret_cast<cl_char *>(ptr) + offset;
     return reinterpret_cast<void *>(byte_ptr);
   }
 
+  /// Given a pointer, create a new USM allocation of the same type
+  ///
+  /// That is, if `old_ptr` points to a host USM allocation, this returns a new
+  /// host allocation.
+  void *sameMemAlloc(void *old_ptr, size_t size, cl_uint alignment,
+                     cl_int *err) {
+    cl_int type_err;
+    cl_unified_shared_memory_type_intel alloc_type = CL_MEM_TYPE_UNKNOWN_INTEL;
+    type_err = clGetMemAllocInfoINTEL(context, old_ptr, CL_MEM_ALLOC_TYPE_INTEL,
+                                      sizeof(alloc_type), &alloc_type, nullptr);
+    EXPECT_SUCCESS(type_err);
+    EXPECT_NE(alloc_type, CL_MEM_TYPE_UNKNOWN_INTEL);
+
+    switch (alloc_type) {
+      case CL_MEM_TYPE_DEVICE_INTEL:
+        return clDeviceMemAllocINTEL(context, device, nullptr, size, alignment,
+                                     err);
+      case CL_MEM_TYPE_HOST_INTEL:
+        return clHostMemAllocINTEL(context, nullptr, size, alignment, err);
+      case CL_MEM_TYPE_SHARED_INTEL:
+        return clSharedMemAllocINTEL(context, device, nullptr, size, alignment,
+                                     err);
+      default:
+        return nullptr;
```

Comment:
Can we ever get here? I would not expect returning `nullptr` to be useful here, but maybe I am not understanding what this default case handles.

Reply:
In theory we shouldn't, but we're in a test suite so anything can happen. I've put an explicit failure here.